### PR TITLE
Improve logging - output formatter selection

### PR DESCRIPTION
**Pull request by kichalla**

### CODE REVIEW

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/MvcCoreLoggerExtensions.cs [hunk 1]
```diff
@@ -368,6 +375,76 @@ static MvcCoreLoggerExtensions()
                 LogLevel.Debug,
                 5,
                 "Current action does not explicitly specify any content types for the response.");
+
+            _selectingOutputFormatterUsingAcceptHeader = LoggerMessage.Define(
+                LogLevel.Debug,
+                6,
+                "Attempting to select an output formatter based on Accept header.");
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/MvcCoreLoggerExtensions.cs]: Should we log the accept header value here?

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/MvcCoreLoggerExtensions.cs [hunk 2]
```diff
@@ -368,6 +375,76 @@ static MvcCoreLoggerExtensions()
...
+
+            _selectingOutputFormatterWithoutUsingAcceptHeader = LoggerMessage.Define(
+                LogLevel.Debug,
+                7,
+                "Attempting to select an output formatter without using Accept header.");
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/MvcCoreLoggerExtensions.cs]: We should also include the content types that were specified on the result so a developer can see why.

File: src/Microsoft.AspNetCore.Mvc.Core/Infrastructure/DefaultOutputFormatterSelector.cs [hunk 3]
```diff
@@ -111,24 +117,33 @@ public override IOutputFormatter SelectFormatter(OutputFormatterCanWriteContext
...
+                    if (!_returnHttpNotAcceptable)
+                    {
+                        selectFormatterWithoutRegardingAcceptHeader = true;
+                    }
                 }
             }
 
             if (selectFormatterWithoutRegardingAcceptHeader)
             {
+                _logger.SelectingOutputFormatterWithoutUsingAcceptHeader();
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/Infrastructure/DefaultOutputFormatterSelector.cs]: If this redundant? Like, we're going to also log @137 and @145 the same info

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/MvcCoreLoggerExtensions.cs (hunk 1 above)
Comment (kichalla) [reply src/Microsoft.AspNetCore.Mvc.Core/Internal/MvcCoreLoggerExtensions.cs]: Ah yes, we should. I didn't realize hosting does not log the accept header as Info where as we do for content-type.

File: src/Microsoft.AspNetCore.Mvc.Core/Infrastructure/DefaultOutputFormatterSelector.cs (hunk 3 above)
Comment (kichalla) [reply src/Microsoft.AspNetCore.Mvc.Core/Infrastructure/DefaultOutputFormatterSelector.cs]: Sure, will remove it.

**Comment by kichalla** [comment]
🆙 📅

File: src/Microsoft.AspNetCore.Mvc.Core/Infrastructure/DefaultOutputFormatterSelector.cs [hunk 4]
```diff
@@ -111,24 +117,31 @@ public override IOutputFormatter SelectFormatter(OutputFormatterCanWriteContext
...
+                        selectFormatterWithoutRegardingAcceptHeader = true;
+                    }
                 }
             }
 
             if (selectFormatterWithoutRegardingAcceptHeader)
             {
                 if (contentTypes.Count == 0)
                 {
+                    _logger.SelectingOutputFormatterWithoutUsingContentTypes();
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/Infrastructure/DefaultOutputFormatterSelector.cs]: It looks to me like this one is also redundant. `SelectFormatterNotUsingContentType` also logs and provides more detail.

**Pull request merged**